README.md
# slowAES (slowAES.decrypt(c, 2, a, b) bypass test cookie
 **slowAES** is a library that allows you to bypass **"test cookie"** used web pages that protected by **aes.js** to prevent web-scrapping/or systems can not run javascripts.

 ## What is slowAES web-scraping protection? ##
 When you get a web page with HttpClient object by nodemcu, your response like below instead of pageview in web browser:

 ```html
<html>
<body>
    <script type="text/javascript" src="/aes.js"></script>
    <script>
        function toNumbers(d) {
            var e = [];
            d.replace(/(..)/g, function(d) {
                e.push(parseInt(d, 16))
            });
            return e
        }

        function toHex() {
            for (var d = [], d = 1 == arguments.length && 
            arguments[0].constructor == Array ? arguments[0] : 
            arguments, e = "", f = 0; f < d.length; f++) e += 
            (16 > d[f] ? "0" : "") + d[f].toString(16);
            return e.toLowerCase()
        }
        var a = toNumbers("f655ba9d09a112d4968c63579db590b4"),
            b = toNumbers("98344c2eee86c3994890592585b49f80"),
            c = toNumbers("a76ce883a42eb6e9fff3e05d9cc6c7e5");
        document.cookie = "__test=" + toHex(slowAES.decrypt(c, 2, a, b)) + 
        "; expires=Thu, 31-Dec-37 23:55:55 GMT; path=/";
        location.href = "http://www.namazvakitleri.site/i=1";
    </script>
    <noscript>
        This site requires Javascript to work, please enable
Javascript in your browser or use a browser with Javascript support
</noscript>
</body>
</html>
```
This means the page you wanted to get has kind of **slowAES protection** based clients ip address. When you get the page with the browser, the code above generate a cookie called **__test** and redirect browser to the page you wanted with that cookie plus a query string i, attemtp etc.

ie.
the page you wanted to get : http://www.namazvakitleri.site

the page seen on browser : http://www.namazvakitleri.site/?i=1 

with a cookie named **__test** and valued : **de1a59425ae4ce0c74d68c19716ffe76**

this means when you try to get this page with HttpClient you have to add a header:

```c
http.addHeader("Cookie", "__test=a6f0d5704838a5ecc1f9fcdd54396a2f", true);
```
and get the page with querystring generated on your page like "**the page seen on browser**" example above.

 
 If you find any problem or bug on this library, please use [Issues](https://github.com/HakkanR/slowAES/issues) feature on github or contact me on my web page 
 
 [devrelerim.com](https://www.devrelerim.com/) **Turkish**
 
 [devrelerim.com](https://en.devrelerim.com/2023/02/slowaes-bypass-test-cookie-nodemcu.html) **English**

 You can follow me on:
 * [Youtube](https://www.youtube.com/HakkanR)
 * [Twitter](https://www.twitter.com/HakkanR)
 * [Instagram](https://www.instagram.com/hakkan_r)
 * [FaceBook](https://www.facebook.com/HakkanR)
 
 ## Basic Usage
 include header file
 ```cpp
 #include <slowAES.h>
 ```
 Get the first page which contains jacascript code, parse your a,b,c values in response and set them to variables called a,b,c: 
 
```c
const char* a = "f655ba9d09a112d4968c63579db590b4";
const char* b = "98344c2eee86c3994890592585b49f80";
const char* c = "a76ce883a42eb6e9fff3e05d9cc6c7e5";
```
Convert them to uint8_t arrays: 
 
```c
uint8_t aNums[16];
uint8_t bNums[16];
uint8_t cNums[16];
toNumbers(a, aNums);
toNumbers(b, bNums);
toNumbers(c, cNums);
```

 Initiate the object and call decrypt **aNums**, **bNums**, **cNums**, arrays:
```c
uint8_t finalCookie[33];
uint8_t resNums[16];
slowAES _slowAES;
_slowAES.decrypt(cNums, aNums, bNums,resNums);
toHex(resNums, finalCookie);
```

Your Final cookie is ready as follows:
```c
String str = (char *)finalCookie;
String fullCookieStr = "__test=" + str;
```

> You can find a working example in examples folder.

## Download
You can download this on
* [Github](https://github.com/HakkanR/slowAES)
* Arduino Library Manager

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

First Release

src/slowAES.h
/**
 * @file slowaes.h
 * @author Hakan ÖZMEN (hakkanr@gmail.com)
 * @brief
 * This library is used for bypassing slowAES protection on webpages can not be got
 * with HttpClient directly. Kind of web pages need a testcookie to pass the page
 * you actually wanted to get. Library generates testcookie based on aes.js.
 * @version 1.0.0
 * @date 2023-02-05
 *
 * @section for more information please visit :
 * https://www.devrelerim.com/ [Turkish]
 * https://en.devrelerim.com/ [English]
 * https://www.youtube.com/HakkanR
 * https://github.com/HakkanR/slowAES
 * 
 * @copyright BDS License, all above text must be included any redistribution.
 *
 */
#ifndef __SLOWAES_H_
#define __SLOWAES_H_

#ifndef _WIN32
#include "Arduino.h"
#endif
class slowAES
{
public:
	slowAES();
	void decrypt(uint8_t *c, uint8_t *a, uint8_t *b, uint8_t *);
private:
	void decrypt(uint8_t *, uint8_t *, uint8_t *);
	void invMain(uint8_t *, uint8_t *, uint8_t);
	void invRound(uint8_t *, uint8_t *);
	void mixColumns(uint8_t *, bool);
	void mixColumn(uint8_t *, bool);
	uint8_t galois_multiplication(uint16_t, uint16_t);
	void subBytes(uint8_t *, bool);
	void shiftRows(uint8_t *, bool);
	void shiftRow(uint8_t *, uint8_t, uint8_t, bool);
	void createRoundKey(uint8_t *, uint8_t, uint8_t *);
	void addRoundKey(uint8_t *, uint8_t *);
	uint8_t numberOfRounds(uint8_t size);
	void expandKey(uint8_t *, uint8_t, uint8_t *);
	void rotate(uint8_t *);
	void core(uint8_t *, uint16_t);

private:
	uint8_t Rcon[255] = {
		0x8d, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8,
		0xab, 0x4d, 0x9a, 0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3,
		0x7d, 0xfa, 0xef, 0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f,
		0x25, 0x4a, 0x94, 0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d,
		0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab,
		0x4d, 0x9a, 0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d,
		0xfa, 0xef, 0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25,
		0x4a, 0x94, 0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d, 0x01,
		0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d,
		0x9a, 0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d, 0xfa,
		0xef, 0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25, 0x4a,
		0x94, 0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d, 0x01, 0x02,
		0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d, 0x9a,
		0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d, 0xfa, 0xef,
		0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25, 0x4a, 0x94,
		0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d, 0x01, 0x02, 0x04,
		0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d, 0x9a, 0x2f,
		0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d, 0xfa, 0xef, 0xc5,
		0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25, 0x4a, 0x94, 0x33,
		0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb};
	uint8_t sbox[256] = {
		0x63, 0x7c, 0x77, 0x7b, 0xf2, 0x6b, 0x6f, 0xc5, 0x30, 0x01, 0x67, 0x2b, 0xfe, 0xd7, 0xab, 0x76,
		0xca, 0x82, 0xc9, 0x7d, 0xfa, 0x59, 0x47, 0xf0, 0xad, 0xd4, 0xa2, 0xaf, 0x9c, 0xa4, 0x72, 0xc0,
		0xb7, 0xfd, 0x93, 0x26, 0x36, 0x3f, 0xf7, 0xcc, 0x34, 0xa5, 0xe5, 0xf1, 0x71, 0xd8, 0x31, 0x15,
		0x04, 0xc7, 0x23, 0xc3, 0x18, 0x96, 0x05, 0x9a, 0x07, 0x12, 0x80, 0xe2, 0xeb, 0x27, 0xb2, 0x75,
		0x09, 0x83, 0x2c, 0x1a, 0x1b, 0x6e, 0x5a, 0xa0, 0x52, 0x3b, 0xd6, 0xb3, 0x29, 0xe3, 0x2f, 0x84,
		0x53, 0xd1, 0x00, 0xed, 0x20, 0xfc, 0xb1, 0x5b, 0x6a, 0xcb, 0xbe, 0x39, 0x4a, 0x4c, 0x58, 0xcf,
		0xd0, 0xef, 0xaa, 0xfb, 0x43, 0x4d, 0x33, 0x85, 0x45, 0xf9, 0x02, 0x7f, 0x50, 0x3c, 0x9f, 0xa8,
		0x51, 0xa3, 0x40, 0x8f, 0x92, 0x9d, 0x38, 0xf5, 0xbc, 0xb6, 0xda, 0x21, 0x10, 0xff, 0xf3, 0xd2,
		0xcd, 0x0c, 0x13, 0xec, 0x5f, 0x97, 0x44, 0x17, 0xc4, 0xa7, 0x7e, 0x3d, 0x64, 0x5d, 0x19, 0x73,
		0x60, 0x81, 0x4f, 0xdc, 0x22, 0x2a, 0x90, 0x88, 0x46, 0xee, 0xb8, 0x14, 0xde, 0x5e, 0x0b, 0xdb,
		0xe0, 0x32, 0x3a, 0x0a, 0x49, 0x06, 0x24, 0x5c, 0xc2, 0xd3, 0xac, 0x62, 0x91, 0x95, 0xe4, 0x79,
		0xe7, 0xc8, 0x37, 0x6d, 0x8d, 0xd5, 0x4e, 0xa9, 0x6c, 0x56, 0xf4, 0xea, 0x65, 0x7a, 0xae, 0x08,
		0xba, 0x78, 0x25, 0x2e, 0x1c, 0xa6, 0xb4, 0xc6, 0xe8, 0xdd, 0x74, 0x1f, 0x4b, 0xbd, 0x8b, 0x8a,
		0x70, 0x3e, 0xb5, 0x66, 0x48, 0x03, 0xf6, 0x0e, 0x61, 0x35, 0x57, 0xb9, 0x86, 0xc1, 0x1d, 0x9e,
		0xe1, 0xf8, 0x98, 0x11, 0x69, 0xd9, 0x8e, 0x94, 0x9b, 0x1e, 0x87, 0xe9, 0xce, 0x55, 0x28, 0xdf,
		0x8c, 0xa1, 0x89, 0x0d, 0xbf, 0xe6, 0x42, 0x68, 0x41, 0x99, 0x2d, 0x0f, 0xb0, 0x54, 0xbb, 0x16};
	// Rijndael Inverted S-box
	uint8_t rsbox[256] = {
		0x52, 0x09, 0x6a, 0xd5, 0x30, 0x36, 0xa5, 0x38, 0xbf, 0x40, 0xa3, 0x9e, 0x81, 0xf3, 0xd7, 0xfb,
		0x7c, 0xe3, 0x39, 0x82, 0x9b, 0x2f, 0xff, 0x87, 0x34, 0x8e, 0x43, 0x44, 0xc4, 0xde, 0xe9, 0xcb,
		0x54, 0x7b, 0x94, 0x32, 0xa6, 0xc2, 0x23, 0x3d, 0xee, 0x4c, 0x95, 0x0b, 0x42, 0xfa, 0xc3, 0x4e,
		0x08, 0x2e, 0xa1, 0x66, 0x28, 0xd9, 0x24, 0xb2, 0x76, 0x5b, 0xa2, 0x49, 0x6d, 0x8b, 0xd1, 0x25,
		0x72, 0xf8, 0xf6, 0x64, 0x86, 0x68, 0x98, 0x16, 0xd4, 0xa4, 0x5c, 0xcc, 0x5d, 0x65, 0xb6, 0x92,
		0x6c, 0x70, 0x48, 0x50, 0xfd, 0xed, 0xb9, 0xda, 0x5e, 0x15, 0x46, 0x57, 0xa7, 0x8d, 0x9d, 0x84,
		0x90, 0xd8, 0xab, 0x00, 0x8c, 0xbc, 0xd3, 0x0a, 0xf7, 0xe4, 0x58, 0x05, 0xb8, 0xb3, 0x45, 0x06,
		0xd0, 0x2c, 0x1e, 0x8f, 0xca, 0x3f, 0x0f, 0x02, 0xc1, 0xaf, 0xbd, 0x03, 0x01, 0x13, 0x8a, 0x6b,
		0x3a, 0x91, 0x11, 0x41, 0x4f, 0x67, 0xdc, 0xea, 0x97, 0xf2, 0xcf, 0xce, 0xf0, 0xb4, 0xe6, 0x73,
		0x96, 0xac, 0x74, 0x22, 0xe7, 0xad, 0x35, 0x85, 0xe2, 0xf9, 0x37, 0xe8, 0x1c, 0x75, 0xdf, 0x6e,
		0x47, 0xf1, 0x1a, 0x71, 0x1d, 0x29, 0xc5, 0x89, 0x6f, 0xb7, 0x62, 0x0e, 0xaa, 0x18, 0xbe, 0x1b,
		0xfc, 0x56, 0x3e, 0x4b, 0xc6, 0xd2, 0x79, 0x20, 0x9a, 0xdb, 0xc0, 0xfe, 0x78, 0xcd, 0x5a, 0xf4,
		0x1f, 0xdd, 0xa8, 0x33, 0x88, 0x07, 0xc7, 0x31, 0xb1, 0x12, 0x10, 0x59, 0x27, 0x80, 0xec, 0x5f,
		0x60, 0x51, 0x7f, 0xa9, 0x19, 0xb5, 0x4a, 0x0d, 0x2d, 0xe5, 0x7a, 0x9f, 0x93, 0xc9, 0x9c, 0xef,
		0xa0, 0xe0, 0x3b, 0x4d, 0xae, 0x2a, 0xf5, 0xb0, 0xc8, 0xeb, 0xbb, 0x3c, 0x83, 0x53, 0x99, 0x61,
		0x17, 0x2b, 0x04, 0x7e, 0xba, 0x77, 0xd6, 0x26, 0xe1, 0x69, 0x14, 0x63, 0x55, 0x21, 0x0c, 0x7d};

private:
	enum modeOfOperation
	{
		OFB,
		CFB,
		CBC
	};
	enum keySize
	{
		SIZE_128 = 16,
		SIZE_192 = 24,
		SIZE_256 = 32
	};
};
void toNumbers(const char *, uint8_t *);
void toHex(uint8_t *arrayPtr, uint8_t *result);

#endif // !__SLOWAES_H_


src/slowAES.cpp
/**
 * @file slowaes.cpp
 * @author Hakan ÖZMEN (hakkanr@gmail.com)
 * @brief This library is used for bypassing slowAES protection on webpages can not be got
 * with HttpClient directly. Kind of web pages need a testcookie to pass the page
 * you actually wanted to get. Library generates testcookie based on aes.js.
 * @version 0.1
 * @date 2023-02-01
 *
 * @section for more information please visit :
 * https://www.devrelerim.com/ [Turkish]
 * https://en.devrelerim.com/ [English]
 * https://www.youtube.com/HakkanR
 * https://github.com/HakkanR/slowAES
 * 
 * @copyright
 * BDS License, all above text must be included any redistribution.
 *
 */
#include "slowAES.h"
slowAES::slowAES()
{
}
void slowAES::decrypt(uint8_t *cipherin, uint8_t *key, uint8_t *iv, uint8_t *bytesOut)
{
	uint8_t output[16];
	decrypt(cipherin, key, output);
	for (int i = 0; i < 16; i++)
		bytesOut[i] = (iv[i] ^ output[i]);
}
void slowAES::decrypt(uint8_t *input, uint8_t *key, uint8_t *output)
{
	uint8_t size = 16;

	uint8_t block[16];
	for (int i = 0; i < 16; i++)
	{
		output[i] = 0;
		block[i] = 0;
	}
	uint8_t nbrRounds = numberOfRounds(size);
	for (uint8_t i = 0; i < 4; i++)		/* iterate over the columns */
		for (uint8_t j = 0; j < 4; j++) /* iterate over the rows */
			block[(i + (j * 4))] = input[(i * 4) + j];
	/* expand the key into an 176, 208, 240 bytes key */
	uint8_t *expandedKey;
	uint8_t expandedKeySize = (16 * (numberOfRounds(size) + 1));
	expandedKey = new uint8_t[expandedKeySize];
	expandKey(key, size, expandedKey);
	/* decrypt the block using the expandedKey */
	invMain(block, expandedKey, nbrRounds);
	for (uint8_t k = 0; k < 4; k++)		/* unmap the block again into the output */
		for (uint8_t l = 0; l < 4; l++) /* iterate over the rows */
			output[(k * 4) + l] = block[(k + (l * 4))];
	delete[] expandedKey;
}
void slowAES::invMain(uint8_t *state, uint8_t *expandedKey, uint8_t nbrRounds)
{
	uint8_t *roundKey = new uint8_t[16];
	createRoundKey(expandedKey, 16 * nbrRounds, roundKey);
	addRoundKey(state, roundKey);

	for (int i = nbrRounds - 1; i > 0; i--)
	{
		createRoundKey(expandedKey, 16 * i, roundKey);
		invRound(state, roundKey);
	}
	shiftRows(state, true);
	subBytes(state, true);
	createRoundKey(expandedKey, 0, roundKey);
	addRoundKey(state, roundKey);
	delete[] roundKey;
}
void slowAES::invRound(uint8_t *state, uint8_t *roundKey)
{
	shiftRows(state, true);
	subBytes(state, true);
	addRoundKey(state, roundKey);
	mixColumns(state, true);
}
void slowAES::mixColumns(uint8_t *state, bool isInv)
{
	uint8_t column[4];
	for (int i = 0; i < 4; i++)
		column[i] = 0;
	/* iterate over the 4 columns */
	for (uint8_t i = 0; i < 4; i++)
	{
		/* construct one column by iterating over the 4 rows */
		for (uint8_t j = 0; j < 4; j++)
			column[j] = state[(j * 4) + i];
		/* apply the mixColumn on one column */
		mixColumn(column, isInv);
		/* put the values back into the state */
		for (uint8_t k = 0; k < 4; k++)
			state[(k * 4) + i] = column[k];
	}
}
void slowAES::mixColumn(uint8_t *column, bool isInv)
{
	uint8_t *mult = new uint8_t[4];
	uint8_t a[] = {14, 9, 13, 11};
	uint8_t b[] = {2, 1, 1, 3};
	if (isInv)
		mult = a;
	else
		mult = b;
	uint8_t cpy[4];
	for (uint8_t i = 0; i < 4; i++)
		cpy[i] = column[i];

	column[0] = galois_multiplication(cpy[0], mult[0]) ^
				galois_multiplication(cpy[3], mult[1]) ^
				galois_multiplication(cpy[2], mult[2]) ^
				galois_multiplication(cpy[1], mult[3]);
	column[1] = galois_multiplication(cpy[1], mult[0]) ^
				galois_multiplication(cpy[0], mult[1]) ^
				galois_multiplication(cpy[3], mult[2]) ^
				galois_multiplication(cpy[2], mult[3]);
	column[2] = galois_multiplication(cpy[2], mult[0]) ^
				galois_multiplication(cpy[1], mult[1]) ^
				galois_multiplication(cpy[0], mult[2]) ^
				galois_multiplication(cpy[3], mult[3]);
	column[3] = galois_multiplication(cpy[3], mult[0]) ^
				galois_multiplication(cpy[2], mult[1]) ^
				galois_multiplication(cpy[1], mult[2]) ^
				galois_multiplication(cpy[0], mult[3]);
}
uint8_t slowAES::galois_multiplication(uint16_t a, uint16_t b)
{
	uint16_t p = 0;
	for (int counter = 0; counter < 8; counter++)
	{
		if ((b & 1) == 1)
			p ^= a;
		if (p > 0x100)
			p ^= 0x100;
		auto hi_bit_set = (a & 0x80); // keep p 8 bit
		a <<= 1;
		if (a > 0x100)
			a ^= 0x100; // keep a 8 bit
		if (hi_bit_set == 0x80)
			a ^= 0x1b;
		if (a > 0x100)
			a ^= 0x100; // keep a 8 bit
		b >>= 1;
		if (b > 0x100)
			b ^= 0x100; // keep b 8 bit
	}
	return (uint8_t)p;
}
void slowAES::subBytes(uint8_t *state, bool isInv)
{
	for (int i = 0; i < 16; i++)
		state[i] = isInv ? rsbox[state[i]] : sbox[state[i]];
}
void slowAES::shiftRows(uint8_t *state, bool isInv)
{
	for (int i = 0; i < 4; i++)
		shiftRow(state, i * 4, i, isInv);
}
void slowAES::shiftRow(uint8_t *state, uint8_t statePointer, uint8_t nbr, bool isInv)
{
	for (uint8_t i = 0; i < nbr; i++)
	{
		if (isInv)
		{
			uint8_t tmp = state[statePointer + 3];
			for (int8_t j = 3; j > 0; j--)
				state[statePointer + j] = state[statePointer + j - 1];
			state[statePointer] = tmp;
		}
		else
		{
			uint8_t tmp = state[statePointer];
			for (uint8_t j = 0; j < 3; j++)
				state[statePointer + j] = state[statePointer + j + 1];
			state[statePointer + 3] = tmp;
		}
	}
}
void slowAES::createRoundKey(uint8_t *expandedKey, uint8_t roundKeyPointer, uint8_t *roundKey)
{
	for (int i = 0; i < 16; i++)
		roundKey[i] = 0;
	for (int i = 0; i < 4; i++)
		for (int j = 0; j < 4; j++)
			roundKey[j * 4 + i] = expandedKey[roundKeyPointer + i * 4 + j];
}
void slowAES::addRoundKey(uint8_t *state, uint8_t *roundKey)
{
	for (uint8_t i = 0; i < 16; i++)
		state[i] ^= roundKey[i];
}
uint8_t slowAES::numberOfRounds(uint8_t size)
{
	uint8_t nbrRounds = 0;
	switch (size)
	{
	case (SIZE_128):
		nbrRounds = 10;
		break;
	case (SIZE_192):
		nbrRounds = 12;
		break;
	case (SIZE_256):
		nbrRounds = 14;
		break;
	}
	return nbrRounds;
}
void slowAES::expandKey(uint8_t *key, uint8_t size, uint8_t *expandedKey)
{
	uint8_t expandedKeySize = (16 * (numberOfRounds(size) + 1));
	uint8_t currentSize = 0;
	uint8_t rconIteration = 1;
	uint8_t t[4] = {0, 0, 0, 0};
	for (uint8_t i = 0; i < expandedKeySize; i++)
		expandedKey[i] = 0;
	/* set the 16,24,32 bytes of the expanded key to the input key */
	for (uint8_t j = 0; j < size; j++)
		expandedKey[j] = key[j];
	currentSize += size;
	while (currentSize < expandedKeySize)
	{
		/* assign the previous 4 bytes to the temporary value t */
		for (uint8_t k = 0; k < 4; k++)
			t[k] = expandedKey[(currentSize - 4) + k];

		/* every 16,24,32 bytes we apply the core schedule to t
		 * and increment rconIteration afterwards
		 */
		if (currentSize % size == 0)
			core(t, rconIteration++);

		/* For 256-bit keys, we add an extra sbox to the calculation */
		if (size == SIZE_256 && ((currentSize % size) == 16))
			for (uint8_t l = 0; l < 4; l++)
				t[l] = sbox[t[l]];

		/* We XOR t with the four-byte block 16,24,32 bytes before the new expanded key.
		 * This becomes the next four bytes in the expanded key.
		 */
		for (uint8_t m = 0; m < 4; m++)
		{
			expandedKey[currentSize] = (expandedKey[currentSize - size] ^ t[m]);
			currentSize++;
		}
	}
}
void slowAES::rotate(uint8_t *word)
{
	uint8_t c = word[0];
	for (uint8_t i = 0; i < 3; i++)
		word[i] = word[i + 1];
	word[3] = c;
}
void slowAES::core(uint8_t *word, uint16_t iteration)
{
	rotate(word);
	for (uint8_t i = 0; i < 4; ++i)
		word[i] = sbox[word[i]];
	/* XOR the output of the rcon operation with i to the first part (leftmost) only */
	word[0] = word[0] ^ Rcon[iteration];
}
void toNumbers(const char *hexStr, uint8_t *arrayPtr)
{
	for (uint8_t i = 0; i < 16; i++)
	{
		uint8_t ch1 = hexStr[i * 2];
		if (ch1 < 58)
			ch1 = ch1 - 48;
		else
			ch1 = ch1 - 87;
		uint8_t ch2 = hexStr[i * 2 + 1];
		if (ch2 < 58)
			ch2 = ch2 - 48;
		else
			ch2 = ch2 - 87;
		arrayPtr[i] = ch1 * 16 + ch2 * 1;
	}
}
void toHex(uint8_t *arrayPtr, uint8_t *result)
{
	for (uint8_t i = 0; i < 16; i++)
	{
		uint8_t ch2 = arrayPtr[i] % 16;
		uint8_t ch1 = (uint8_t)((arrayPtr[i] - ch2) / 16);
		if (ch2 < 10)
			result[i * 2 + 1] = (char)ch2 + 48;
		else
			result[i * 2 + 1] = (char)ch2 + 87;
		if (ch1 < 10)
			result[i * 2] = (char)ch1 + 48;
		else
			result[i * 2] = (char)ch1 + 87;
	}
	result[32] = '\0';
}